File: ruby/io_benchmark.rb
require 'thread'
require 'fileutils'

def create_test_files(count, task_id)
  errors = []
  threads = []
  mutex = Mutex.new
  
  count.times do |i|
    threads << Thread.new(i) do |index|
      begin
        content = "Test file #{index} content with some data: #{rand}"
        filename = "temp_#{task_id}_#{index}.txt"
        File.write(filename, content)
      rescue => e
        mutex.synchronize { errors << e }
      end
    end
  end
  
  threads.each(&:join)
  
  raise errors.first unless errors.empty?
end

def read_test_files(count, task_id)
  results = Array.new(count)
  errors = []
  threads = []
  mutex = Mutex.new
  
  count.times do |i|
    threads << Thread.new(i) do |index|
      begin
        filename = "temp_#{task_id}_#{index}.txt"
        # リトライロジック
        10.times do |retry_count|
          begin
            results[index] = File.read(filename)
            break
          rescue Errno::ENOENT => e
            raise e if retry_count == 9
            sleep 0.01
          end
        end
      rescue => e
        mutex.synchronize { errors << e }
      end
    end
  end
  
  threads.each(&:join)
  
  raise errors.first unless errors.empty?
  
  results
end

def cleanup_test_files(count, task_id)
  threads = []
  
  count.times do |i|
    threads << Thread.new(i) do |index|
      begin
        filename = "temp_#{task_id}_#{index}.txt"
        File.delete(filename)
      rescue
        # Ignore errors
      end
    end
  end
  
  threads.each(&:join)
end

def simulate_network_delay(ms)
  sleep(ms / 1000.0)
end

def io_intensive_task(task_id)
  file_count = 50
  network_calls = 20
  
  # Create files concurrently
  create_test_files(file_count, task_id)
  
  # Simulate network calls with delays
  network_threads = []
  network_calls.times do
    delay = 10 + rand(20)
    network_threads << Thread.new { simulate_network_delay(delay) }
  end
  
  # Read files concurrently while network calls are happening
  file_contents = read_test_files(file_count, task_id)
  
  # Wait for network calls to complete
  network_threads.each(&:join)
  
  # Cleanup
  cleanup_test_files(file_count, task_id)
  
  file_contents.length
end

def run_io_benchmark(task_count)
  puts "Ruby I/O benchmark start"
  
  start_time = Time.now
  
  task_threads = []
  task_count.times do |i|
    task_threads << Thread.new(i) do |index|
      begin
        task_id = "task#{index}"
        io_intensive_task(task_id)
        puts "Task #{index+1} done"
      rescue => e
        puts "Task #{index+1} failed: #{e}"
      end
    end
  end
  
  task_threads.each(&:join)
  
  elapsed = Time.now - start_time
  puts "All #{task_count} tasks done in #{elapsed.round(4)}s"
end

File: python/io_benchmark.py
import os
import random
import time
import threading
from queue import Queue

def create_test_files(count):
    errors = Queue()
    
    def create_file(i):
        try:
            content = f"Test file {i} content with some data: {random.random()}"
            filename = f"temp_{i}.txt"
            with open(filename, 'w') as f:
                f.write(content)
        except Exception as e:
            errors.put(e)
    
    threads = []
    for i in range(count):
        thread = threading.Thread(target=create_file, args=(i,))
        thread.start()
        threads.append(thread)
    
    for thread in threads:
        thread.join()
    
    if not errors.empty():
        raise errors.get()

def read_test_files(count):
    results = [None] * count
    errors = Queue()
    
    def read_file(i):
        try:
            filename = f"temp_{i}.txt"
            with open(filename, 'r') as f:
                results[i] = f.read()
        except Exception as e:
            errors.put(e)
    
    threads = []
    for i in range(count):
        thread = threading.Thread(target=read_file, args=(i,))
        thread.start()
        threads.append(thread)
    
    for thread in threads:
        thread.join()
    
    if not errors.empty():
        raise errors.get()
    
    return results

def cleanup_test_files(count):
    def cleanup_file(i):
        try:
            filename = f"temp_{i}.txt"
            os.remove(filename)
        except:
            pass  # Ignore errors
    
    threads = []
    for i in range(count):
        thread = threading.Thread(target=cleanup_file, args=(i,))
        thread.start()
        threads.append(thread)
    
    for thread in threads:
        thread.join()

def simulate_network_delay(ms):
    time.sleep(ms / 1000.0)

def io_intensive_task():
    file_count = 50
    network_calls = 20
    
    # Create files concurrently
    create_test_files(file_count)
    
    # Simulate network calls with delays
    network_threads = []
    for _ in range(network_calls):
        delay = 10 + random.randint(0, 19)
        thread = threading.Thread(target=simulate_network_delay, args=(delay,))
        thread.start()
        network_threads.append(thread)
    
    # Read files concurrently while network calls are happening
    file_contents = read_test_files(file_count)
    
    # Wait for network calls to complete
    for thread in network_threads:
        thread.join()
    
    # Cleanup
    cleanup_test_files(file_count)
    
    return len(file_contents)

def run_io_benchmark(task_count):
    print("Python I/O benchmark start")
    
    start = time.time()
    
    def io_task(i):
        try:
            io_intensive_task()
            print(f"Task {i+1} done")
        except Exception as e:
            print(f"Task {i+1} failed: {e}")
    
    threads = []
    for i in range(task_count):
        thread = threading.Thread(target=io_task, args=(i,))
        thread.start()
        threads.append(thread)
    
    for thread in threads:
        thread.join()
    
    elapsed = time.time() - start
    print(f"All {task_count} tasks done in {elapsed:.4f}s")